addons/mail/tests/test_ir_mail_server.py

```diff
@@ -441,3 +444,253 @@ def test_personal_mail_server(self):
+        )
+        delayed = self.env["mail.mail"].search(
+            [('id', 'in', new_mails_user_1.ids), ('state', '!=', 'sent')])
+        self.assertEqual(len(delayed), 50)
+        scheduled_dates = delayed.mapped('scheduled_date')
+        self.assertTrue(all(scheduled_dates))
+        scheduled_dates = sorted(scheduled_dates)
+        self.assertEqual(set(scheduled_dates[:20]), {datetime(2025, 1, 1, 20, 6, 23)})
+        self.assertEqual(set(scheduled_dates[20:40]), {datetime(2025, 1, 1, 20, 7, 23)})
+        self.assertEqual(set(scheduled_dates[40:]), {datetime(2025, 1, 1, 20, 8, 23)})
+
+        with self.mock_smtplib_connection(), self._freeze_time("2025-01-01 20:06:23"):
+            self.env['mail.mail'].process_email_queue()
+
+        self.assertEqual(mail_server_1.owner_limit_count, 20)
+        self.assertEqual(mail_server_1.owner_limit_time, datetime(2025, 1, 1, 20, 6, 23))
+
+        # Should send in priority the old mails
+        mails_sent = self.env["mail.mail"].search(
+            [('id', 'in', mails_user_1.ids), ('state', '=', 'sent')])
+        self.assertEqual(len(mails_sent), 40, "20 more emails are sent")
+        self.assertEqual(mails_sent, mails_user_1[:40])
+        self.assertEqual(set(new_mails_user_1.mapped('state')), {'outgoing'})
+
+        # The CRON re-run in the same minute, skip the sending
+        with self.mock_smtplib_connection(), self._freeze_time("2025-01-01 20:06:23"):
+            self.env['mail.mail'].process_email_queue()
+        self.assertEqual(self.env["mail.mail"].search_count(
+            [('id', 'in', mails_user_1.ids), ('state', '=', 'sent')]),
+            40,
+            "No more emails sent",
+        )
+
+        # The CRON run in one minute later, we can 20 more emails
+        with self.mock_smtplib_connection(), self._freeze_time("2025-01-01 20:07:23"):
+            self.env['mail.mail'].process_email_queue()
+
+        self.assertEqual(mail_server_1.owner_limit_count, 20)
+        self.assertEqual(mail_server_1.owner_limit_time, datetime(2025, 1, 1, 20, 7, 23))
+
+        self.assertEqual(self.env["mail.mail"].search_count(
+            [('id', 'in', mails_user_1.ids), ('state', '=', 'sent')]),
+            60,
+            "20 more emails are sent",
+        )
+
+        # The CRON is late compared to the scheduled mails,
+        # it should re-schedule the mails, starting from the current time
+        with self.mock_smtplib_connection(), self._freeze_time("2025-01-01 20:25:23"):
+            self.env['mail.mail'].process_email_queue()
+
+        self.assertEqual(mail_server_1.owner_limit_count, 20)
+        self.assertEqual(mail_server_1.owner_limit_time, datetime(2025, 1, 1, 20, 25, 23))
+
+        self.assertEqual(self.env["mail.mail"].search_count(
+            [('id', 'in', mails_user_1.ids), ('state', '=', 'sent')]),
+            80,
+            "20 more emails are sent",
+        )
+
+        # Should send in priority the old mails
```

Comment:
also check if we can split some part of the test into other unit tests. Big tests are a pain to maintain, as if something changes at the beginning, everything else changes. Unit tests should be small and check a given feature, not be lengthy tests you don't understand one month after having finished it ^^